File: cowichan/baSrcPaper/go/chain/seq/src/all/winnow.go
/*
 * winnow: weighted point selection
 *
 * input:
 *   Randmat_matrix: an integer matrix, whose values are used as masses
 *   Thresh_mask: a boolean matrix showing which points are eligible for
 *     consideration
 *   nrows, ncols: the number of rows and columns
 *   nelts: the number of points to select
 *
 * output:
 *   Winnow_points: a vector of (x, y) points
 *
 */
package all

import (
  "sort"
  "flag"
)

var Is_bench = flag.Bool("is_bench", false, "")

type WinnowPoint struct {
  value byte;
  i, j int;
}

var Winnow_points []WinnowPoint;

type WinnowPoints []WinnowPoint;

func (p WinnowPoints) Len() int { return len(p) }
func (p WinnowPoints) Swap(i, j int) { p[i], p[j] = p[j], p[i] }
func (p WinnowPoints) Less(i, j int) bool {
  if p[i].value < p[j].value {
    return true;
  }
  if p[i].value > p[j].value {
    return false;
  }
  if p[i].i < p[j].i {
    return true;
  }
  if p[i].i > p[j].i {
    return false;
  }
  return p[i].j < p[j].j;
}

func Winnow(nrows, ncols, nelts int) {
  Winnow_points = make ([]WinnowPoint, nelts)
  var n = 0;

  for i := 0; i < nrows; i++ {
    for j := 0; j < ncols; j++ {
      if (*Is_bench) {
        if (((i * j) % (ncols + 1)) == 1) {
          Thresh_mask[i][j] = 1;
        }
      }
      if Thresh_mask[i][j] == 1 {
        n++;
      }
    }
  }

  values := make(WinnowPoints, n);

  var count = 0;
  for i := 0; i < nrows; i++ {
    for j := 0; j < ncols; j++ {
      if Thresh_mask[i][j] == 1 {
        values[count] = WinnowPoint{Randmat_matrix[i][j], i, j};
        count++;
      }
    }
  }
  
  sort.Sort(values);

  var total = len(values);
  var chunk int = total / nelts;

  for i := 0; i < nelts; i++ {
    var index = i * chunk;
    Winnow_points[i] = values[index];
  }
}

File: cowichan/baSrcPaper/tbb/chain/seq/winnow.cc
/* winnow: weighted point selection
 *
 * input:
 *   randmat_matrix: an integer matrix, whose values are used as masses
 *   thresh_mask: a boolean matrix, showing which points are eligible for
 *     consideration
 *   nrows, ncols: number of rows and columns
 *   nelts: the number of points to select
 *
 * output:
 *   winnow_points: a vector of (x, y) points
 */
#include <cassert>
#include <cstdio>
#include <cstdlib>
#include <cstring>

#include <algorithm>
#include <iostream>
#include <vector>

using namespace std;

extern int is_bench;
extern int *randmat_matrix;
extern int *thresh_mask;
pair<int, int> *winnow_points;
static pair<int, pair<int, int> > *values;

void winnow(int nrows, int ncols, int nelts) {
  winnow_points = (pair<int,int> *) malloc (sizeof (pair<int,int>) * nelts);
  values = (pair<int, pair<int,int> > *) malloc (sizeof (pair<int, pair<int,int> >) * nrows  * ncols);

  int count = 0;
  for (int i = 0; i < nrows; i++) {
    for (int j = 0; j < ncols; j++) {
      if (thresh_mask[i*ncols + j]) {
        values[count++] = (make_pair(randmat_matrix[i*ncols + j], make_pair(i, j)));
      }
    }
  }

  sort(values, values + count);

  size_t n = count;
  size_t chunk = n / nelts;

  for (int i = 0; i < nelts; i++) {
    int index = i * chunk;
    winnow_points[i] = values[index].second;
  }
  free (values);
}
